packages/styled-components/src/native/test/native.test.js

```diff
@@ -446,4 +446,45 @@ Object {
       expect(() => wrapper.root.findByType('Text')).toThrowError();
     });
   });
+
+  describe('extended CSS support', () => {
+    it('should handle vh, vw, vmin, vmax and rem units', () => {
+      const Comp = styled(View)`
+        width: 10vmin;
+        height: 10vmax;
+        padding: 10vw 10vh;
+        border: 1rem solid black;
+      `;
```

Comment:
I'm not sure what's the policy with unit tests in styled-components, but these two tests that you have don't properly cover the changes that you have made in the code.

What about adding tests for `vmin`/`vmax`, for the `Dimensions` event listener, for `rem` units, for different CSS property combinations, for invalid code, etc.?

Reply:
oh sorry, vmin/vmax is already tested.

Reply:
Good remark. I'm not sure about how to emulate screen resize, but I'll try. rem is tested already in the border width. What should I do in case of invalid code?

Reply:
Sorry, I don't find how to emulate a resize of the view...

Reply:
About bad inputs, it seems that the current codebase doesn't handle them, so I don't think that it should be part of this PR...

Reply:
> What should I do in case of invalid code?

I was thinking of some tests that verify that your code path only runs for certain properties and gets skipped with a property that is not supported. React Native will throw an error if you try to give it invalid styling, so I'm not sure if it actually would make any sense to test that.

Reply:
> Sorry, I don't find how to emulate a resize of the screen or an orientation change...

I don't remember how to do it exactly, but I guess that you would need to mock the functions in the `Dimensions` module, trigger the event listener and then re-render the component. I'm not sure if there is some other way to do that easier.

Reply:
> > What should I do in case of invalid code?
> 
> I was thinking of some tests that verify that your code path only runs for certain properties and gets skipped with a property that is not supported.
> 
> React Native will throw an error if you try to give it invalid styling, so I'm not sure if it actually would make any sense to test that.

I tried

```
       const Comp = styled(View)`
        width: 10vwf;
        padding: 10vwqsd 10fdsvw;
        border: 10vw soli black;
        heigsh: 10vw;
      `;
```

But I get :

```
Failed to parse declaration "padding: 10vwqsd 10fdsvw"

      137 |         // support, so we'll just disable multiple values here.
      138 |         // https://github.com/styled-components/css-to-react-native/issues/11
    > 139 |         const styleObject = transformDeclPairs(declPairs, [
```

Which is not my doing...

Reply:
> > Sorry, I don't find how to emulate a resize of the screen or an orientation change...
> 
> I don't remember how to do it exactly, but I guess that you would need to mock the functions in the `Dimensions` module, trigger the event listener with updated screen size and then re-render the component. I'm not sure if there is some other way to do that easier.

I don't think that I can manually fire an event on Dimensions listeners. Check [the code](https://github.com/facebook/react-native/blob/bcc62a6d179e673754a46f9d93ab61dbe0576894/Libraries/Utilities/Dimensions.js) . I have no way to access the eventEmitter...

Reply:
I'll try something. I have an idea

Reply:
Nope. Didn't work. This code:
```
  React.useEffect(() => {
    Dimensions.addEventListener('change', sizeListener)
    return () => Dimensions.removeEventListener('change', sizeListener)
  }, [])
```
is executed only after the tests are over.
I don't see how I can make the tests wait for this to happen...

Reply:
> I tried

As I said previously, it does not make sense to test invalid CSS, because the CSS parser or React Native will throw an error anyway.

What could make sense would be to test that the properties that can have those vm/vh/vmin/vmax/em units are actually going through the code path that you added, but as I have not been working with the styled-components, I'm not sure if those kind of things are usually covered with tests.

Reply:
> I don't think that I can manually fire an event on Dimensions listeners.

What if you would just mock the whole `Dimensions` API? Then you shouldn't have to worry about actually dispatching events.

...and looking at the test renderer docs, it seems that you can mount a component and the just update it after: https://reactjs.org/docs/test-renderer.html

Reply:
That was my idea. This is what I tried:

```javascript
    it('should handle window resizing', () => {
      // Create the component to be tested
      const Comp = styled(View)`
        width: 10vw;
        padding: 10vw 10vw;
        border: 10vw solid black;
      `;

      // Getting current window dimension
      const { width, height } = Dimensions.get('window')
      const vw = width / 10

      // We want to restore the Dimensions object after our test
      const oldDimensions = {...Dimensions}

      // Mock the Dimensions object
      const listeners = []
      Dimensions.addEventListener = (type, listener) => type==='change' && listeners.push(listener)
      Dimensions.removeEventListener = (type, listener) => type==='change' && listeners.splice(listeners.indexOf(listener), 1)
      Dimensions.get = () => ({ width, height })

      // Initial render of the View
      const wrapper = TestRenderer.create(<Comp />);
      const view = wrapper.root.findByType('View');
      expect(view.props.style).toEqual([{
        width: vw,
        paddingTop: vw, paddingBottom: vw, paddingLeft: vw, paddingRight: vw,
        borderColor: 'black', borderStyle: 'solid', borderWidth: vw
      }]);

      // Emulate a resize
      Dimensions.get = () => ({ width: 200, height })
      listeners.forEach(listener => listener())

      // Trying to update the view
      wrapper.update()
      const updatedView = wrapper.root.findByType('View');
      // /!\ The line above generate the error: Can't access .root on unmounted test renderer
      expect(updatedView.props.style).toEqual([{
        width: 20,
        paddingTop: 20, paddingBottom: 20, paddingLeft: 20, paddingRight: 20,
        borderColor: 'black', borderStyle: 'solid', borderWidth: 20
      }]);

      // Restore the old Dimensions object.
      Object.assign(Dimensions, oldDimensions)
    })
```

I get `Can't access .root on unmounted test renderer`.

Am I doing something wrong?

Reply:
Ok @kristerkari I got it fixed after hours of bumping my head on the wall. react-test-renderer v16.9 let you use await act(async () => {}) to wait for useEffects to happen. In the end I could successfully emulate a resize and confirm that the code is working.